README.md
# KZVG: Заполнятор договоров 3000

Заполняет шаблон договора в .docx данными из формы. Помогает не множить сущности и не искать по коллегам "последний использованный договор, чтобы скопировать туда данные нового клиента": правильный шаблон всегда ведется в одном месте, а контракты только происходят от него, наследуя общие признаки.

Собрано на коленке для нужд юридической фирмы [Казарновски Групп](https://kzvg.ru), в которой я руковожу таможенной практикой. Пока что работает только под Microsoft Windows.

Архивы с бинарниками - [в разделе Releases](https://github.com/medotkato/docbuilder/releases). Скачиваем, распаковываем, читаем README. Запускаем и работаем.

## Порядок использования:

1. Размечаешь .docx шаблон в /in/contract_template.docx, вставляя в него плейсхолдеры вида {{ some_placeholder1 }} ... {{ some_placeholder2 }} там, где надо заполнять данными из формы
2. Создаешь конфиг для формы в /in/contract_config.yaml, в котором указывашь заголовок формы, лого и поля для заполнения (some_placeholder1, some_placeholder2, etc.) - это yaml, там все просто и очевидно.
3. Запускаешь форму

    ``` bash
    python formbuilder.py -c config.yaml
    ```

    Ну или просто

    ``` bash
    test\formbuilder.cmd
    ```

4. Заполняешь поля формы, как надо, и жмешь "Заполнить договор"
5. Забираешь заполненный договор в папке /out/docx с именем вида YYMMDD-HHMMSS_document_filled.docx
6. ???
7. PROFIT!

## ETC

- История разработки: [Changelog.md](CHANGELOG.md)
- Что еще нужно (хотелось бы) сделать: [TODO.md](TODO.md)
- Лицензия: Копилефт, GPL v3 и CC BY-SA 4.0. См. [License.md](LICENCE.md)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [Unreleased]

-

## [0.66] - 2021-06-25

### Added

- build.cmd для сборки приложения pyinstaller-ом
- config.yaml в корне проекта для хранения параметров всего приложения: где забирать шаблоны, куда складывать результат, где хранятся картинки (помогает быстро перестроить приложение под другие нужды), где лежит дефолтный конфиг документа для запуска приложения

### Changed

- теперь приложение собирается из 2 конфигов: общие настройки из config.yaml, настройки документа (и формы) из contract_config.yaml, что облегчает кастомизацию приложения под будущие нужды и позволяет в будущем делать hot-plug конфига с документом
- в верхней части формы выводится не название приложения (которое теперь осталось только в заголовке окна), а название заполняемого документа, которое подтягивается из конфига документа (например, contract_config.yaml)
- теперь шаблон документа называется contract_template.docx, а его конфиг - contract_config.yaml)

## [0.55] - 2021-06-24

### Added

- протестировал упаковку в бинарник с помощью pyinstall
- оформил requirements.txt для быстрого развертывания нужной для исполнения среды с pip:

  ```
  pip install -r requirements.txt
  ```

## [0.5] - 2021-06-23

### Added

- заполнение docx документа происходит по нажатию кнопке "Заполнить договор"
- все собрано воедино и протестировано

### Changed

- шаблоны документа и формы теперь лежат в /in
- выходные документы складываются в /out/data_yaml (сырые данные) и /out/docx (документ Word)
- документ и форма собираются из конфига в кодировке UTF-8, а не CP-1251


## [0.2] - 2021-06-22

### Added

- readme.md с ответами на вопросы "зачем" и "как"
- метки полей в yaml с данными документа
- отображение формы wx на основе конфига/шаблона в .yaml (formbuilder.py)
- файлы лицензий и changelog
- formbuilder для создания формы заполнения договора
- тестовые скрипты для запуска докбилдера и формы
- Some_Contract_Template.docx с открытым шаблоном контракта KZVG, но без некоторых чувствительных деталей
- аргументы в docbuilder и frombuilder для CLI

### Changed

- создание документа из .docx шаблона теперь из конфига в .yaml, а не json (docbuilder.py)
- раздение кода docbuiler на yaml_handler и docbuilder

## [0.1] - 2021-05-05

- docbuilder.py, который заполняет .docx шаблон с плейсхолдерами данными из json-конфига


src/wxtest.py
import wx
from docbuilder import *

class FormBuilder(wx.Frame):

    def __init__(self, parent, title, logo, fields):
        super(FormBuilder, self).__init__(parent, title=title, size=(500, 800))

        self.InitUI(title, logo, fields)
        self.Centre()

    def InitUI(self, title, logo, fields):

        # panel = wx.Panel(self)

        # Add a panel so it looks the correct on all platforms
        panel = wx.ScrolledWindow(self,wx.ID_ANY)
        panel.SetScrollbars(0, 1, 0, 0)

        sizer = wx.GridBagSizer(15,15)

        icon = wx.StaticBitmap(panel, bitmap=wx.Bitmap(logo))
        sizer.Add(icon, pos=(0, 0), flag=wx.TOP|wx.LEFT|wx.ALIGN_LEFT, border=5)

        text1 = wx.StaticText(panel, label=title)
        sizer.Add(text1, pos=(0, 2), flag=wx.TOP|wx.RIGHT, border=15)

        line = wx.StaticLine(panel)
        sizer.Add(line, pos=(1, 0), span=(1, 5),
        flag=wx.EXPAND|wx.BOTTOM, border=10)

        i = 2
        for field in fields.items():
            text = wx.StaticText(panel, label=field[1]['label'])
            sizer.Add(text, pos=(i, 0), flag=wx.LEFT, border=10)
            tc = wx.TextCtrl(panel, value=str(field[1]['value']), name=field[0])
            sizer.Add(tc, pos=(i, 1), span=(1, 3), flag=wx.TOP|wx.EXPAND)
            i = i+1

        button = wx.Button(panel, label="Заполнить договор", size=(500,30))
        sizer.Add(button, pos=(i, 1), span=(1, 2), flag=wx.BOTTOM|wx.RIGHT, border=10)

        sizer.AddGrowableCol(2)

        panel.SetSizer(sizer)
        sizer.Fit(self)


def main():

    yaml_file = 'doc_details.yaml'
    details = yaml_reader(yaml_file)

    form_name = details['form_name']
    form_logo_png = details['form_logo_png']
    form_fields = details['form_fields']

    app = wx.App()
    ex = FormBuilder(None, title=form_name, logo=form_logo_png, fields=form_fields)
    ex.Show()
    app.MainLoop()

if __name__ == '__main__':
    main()


src/docbuilder.py
import argparse
from yaml_handler import *

def values_extractor (yaml_full: dict) -> dict:
    dic = dict()
    for field in yaml_full.items():
        key = field[0]
        val=str(field[1]['value'])
        dic.update({key: val})
    return dic

def docbuilder (template_docx_filename, doc_data, out_docx_filename):
    from docxtpl import DocxTemplate
    doc = DocxTemplate(f'{template_docx_filename}')
    doc.render(doc_data)
    print (f'Writing to: {out_docx_filename}')
    doc.save(f'{out_docx_filename}')

def main():

    my_parser = argparse.ArgumentParser(description='Fill the .docx template with data from .yaml')

    my_parser.add_argument('-t', '--template',
                        metavar='some_template.docx',
                        type=str,
                        help='docx template\'s filename')
    my_parser.add_argument('-d', '--data',
                        metavar='data.yaml',
                        type=str,
                        help='data for filling the template in')
    my_parser.add_argument('-o', '--outfile',
                    metavar='outfile.docx',
                    type=str,
                    help='filename for storing the result')

    args = my_parser.parse_args()

    template_docx_filename = args.template
    data_yaml_filename = args.data
    out_docx_filename = args.outfile

    print (f'Creating {out_docx_filename} using data from {data_yaml_filename} and the .docx template from {template_docx_filename}...')

    details_full = yaml_read (data_yaml_filename)
    details_short = values_extractor (details_full)
    docbuilder (template_docx_filename, details_short, out_docx_filename)

if __name__ == '__main__':
    main()


src/yaml_handler.py
import yaml

def yaml_read (yaml_file: str) -> dict:
    # scriptfolder = os.path.dirname(os.path.abspath(__file__))
    # print (f'scriptfolder is {scriptfolder}')
    try:
        with open(f'{yaml_file}','r', encoding="utf-8") as yaml_file:
            details = yaml.safe_load(yaml_file)
    except:
        raise Exception(f'Can\'t find ({yaml_file}) or can\'t load it.')
    return details

def yaml_write (yaml_file: str, yaml_data: dict):
    try:
        with open(f'{yaml_file}','w', encoding="utf-8") as yaml_file:
            yaml.dump(yaml_data, yaml_file, allow_unicode=True, explicit_start=True, indent=4, sort_keys=False)
    except:
        raise Exception(f'Can\'t find ({yaml_file}) or can\'t write to it.')
    return 1

def main():
    return 1

if __name__ == '__main__':
    main()


src/formbuilder.py
import wx
import yaml_handler
import docbuilder
import argparse
import re
from datetime import datetime

class FormBuilderApp (wx.App):

    def __init__ (self, app_config, doc_config):

        super(FormBuilderApp, self).__init__()

        self.frame = FormBuilderFrame(None, app_config, doc_config)
        self.SetTopWindow(self.frame)
        self.frame.Show()

class FormBuilderFrame(wx.Frame):

    def __init__(self, parent, app_config, doc_config):
        app_title = app_config['app_name']['value']
        super(FormBuilderFrame, self).__init__(parent, title=app_title, size=(500, 800))
        self.InitUI(app_config, doc_config) # create UI
        self.Centre() # put the form in the center of the screen

    def InitUI(self, app_config, doc_config):

        doc_name = doc_config['doc_title']['value']
        doc_fields = doc_config['doc_fields']['value']
        app_logo_path = app_config['res_folder']['value'] + app_config['app_logo']['value']
        button_label = doc_config['form_button_label']['value']

        # Add a scrolled self.panel so it looks the correct on all platforms
        self.panel = wx.ScrolledWindow(self,wx.ID_ANY)
        self.panel.SetScrollbars(0, 1, 0, 0)

        sizer = wx.GridBagSizer(10,10)
        i = 0

        icon = wx.StaticBitmap(self.panel, bitmap=wx.Bitmap(app_logo_path))
        sizer.Add(icon, pos=(i, 0), flag=wx.TOP|wx.LEFT|wx.ALIGN_LEFT, border=5)
        text1 = wx.StaticText(self.panel, label=doc_name)
        sizer.Add(text1, pos=(i, 1), flag=wx.TOP|wx.LEFT, border=15)
        i = i + 1

        line = wx.StaticLine(self.panel)
        sizer.Add(line, pos=(1, 0), span=(1, 5),
        flag=wx.EXPAND|wx.BOTTOM, border=10)
        i = i + 1

        for field in doc_fields.items():
            field_label = field[1]['label']
            field_value = field[1]['value']
            field_name = field[0]
            tc_name = f"{field_name}::{field_label}"

            text = wx.StaticText(self.panel, label=field_label)
            sizer.Add(text, pos=(i, 0), flag=wx.LEFT|wx.ALIGN_RIGHT, border=10)

            tc = wx.TextCtrl(self.panel, value=str(field_value), name=tc_name)
            sizer.Add(tc, pos=(i, 1), span=(1, 3), flag=wx.TOP|wx.EXPAND)
            i = i + 1

        button = wx.Button(self.panel, label=button_label, size=(500,30))
        sizer.Add(button, pos=(i, 1), span=(1, 2), flag=wx.BOTTOM|wx.RIGHT, border=10)
        self.Bind(wx.EVT_BUTTON, lambda event: self.OnButton_Fill_Click (event, app_config, doc_config), id=button.GetId())

        sizer.AddGrowableCol(2)

        self.panel.SetSizer(sizer)
        sizer.Fit(self)

    def grab_fields_data(self) -> dict:
        data_out_dic = dict()
        sizer = self.panel.GetSizer()
        children = sizer.GetChildren()
        for child in children:
            widget = child.GetWindow()
            if isinstance(widget, wx.TextCtrl):
                field_name, field_label = re.split('::', widget.GetName())
                field_value = widget.GetValue()
                data_out_dic.update({field_name: {'label': field_label, 'value': field_value}})
        return data_out_dic

    def OnButton_Fill_Click (self, event, app_config, doc_config):


        data_out_dic = self.grab_fields_data()
        docx_data_out_dic = docbuilder.values_extractor(data_out_dic)

        now = datetime.now()
        dt_string = now.strftime("%y%m%d-%H%M%S")

        out_folder_data = app_config['out_folder_data']['value']
        out_folder_docx = app_config['out_folder_docx']['value']
        docx_template_path = app_config['in_folder']['value'] + doc_config['doc_template']['value']

        data_out_path = f'{out_folder_data}{dt_string}_data.yaml'
        docx_out_path = f'{out_folder_docx}{dt_string}_document_filled.docx'

        yaml_handler.yaml_write(data_out_path, data_out_dic)
        docbuilder.docbuilder(docx_template_path, docx_data_out_dic, docx_out_path)

        return True


def main():

    my_parser = argparse.ArgumentParser(description='Fill the .docx template with data from .yaml')

    my_parser.add_argument('-c', '--config',
                        metavar='config.yaml',
                        type=str,
                        help='application\'s config in .yaml file')

    args = my_parser.parse_args()

    app_config_path = args.config

    print (f'Starting app with config from: {app_config_path}')
    app_config = yaml_handler.yaml_read(app_config_path)

    doc_config_path = app_config['in_folder']['value'] + app_config['doc_config_default']['value']

    print (f'Building form using template from: {doc_config_path}')

    doc_config = yaml_handler.yaml_read(doc_config_path)

    app = FormBuilderApp(app_config, doc_config)
    app.MainLoop()

if __name__ == '__main__':
    main()
